File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/interactors/MainEntityListInteractor.java
package com.example.rafaelanastacioalves.moby.domain.interactors;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;

import com.example.rafaelanastacioalves.moby.domain.entities.MainEntity;
import com.example.rafaelanastacioalves.moby.domain.entities.Resource;
import com.example.rafaelanastacioalves.moby.retrofit.AppRepository;

import java.util.List;

import javax.inject.Inject;

public class MainEntityListInteractor implements Interactor<MainEntityListInteractor.RequestValues>{

    private final AppRepository appRepository;

    @Inject
    MainEntityListInteractor(AppRepository appRepository){
        this.appRepository = appRepository;
    }
    @Override
    public LiveData<Resource<List<MainEntity>>> execute(RequestValues requestValues) {
        MutableLiveData<Resource<List<MainEntity>>> repositoryLiveData = (MutableLiveData<Resource<List<MainEntity>>>) appRepository.getMainEntityList();
        // aqui podemos aplicar transformações baseadas em regras de negócio usando
        // Transformations. Ex.: Transformations.map()
        Transformations.map(repositoryLiveData, resource -> {

            switch (resource.status){
                case GENERIC_ERROR:
                    resource.message = "Main entity list error";
            }

            return resource;

        });
        return repositoryLiveData ;
    }

    public static class RequestValues implements Interactor.RequestValues {
        // in this case we don't need nothing for this use case
    }
}

File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/interactors/MainEntityListInteractor.kt
package com.example.rafaelanastacioalves.moby.domain.interactors

import com.example.rafaelanastacioalves.moby.domain.entities.MainEntity
import com.example.rafaelanastacioalves.moby.domain.entities.Resource
import com.example.rafaelanastacioalves.moby.retrofit.AppRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext

class MainEntityListInteractor :
        Interactor<Resource<List<MainEntity>>, MainEntityListInteractor.RequestValues>() {

    val appRepository: AppRepository


    init {
        appRepository = AppRepository
    }

    override suspend fun run(requestValues: RequestValues?): Resource<List<MainEntity>> {
        var finalList: List<MainEntity> = ArrayList<MainEntity>()

        withContext(Dispatchers.IO) {

            // in this examaple we could call sequentially or wait for one result so we get some data to make another call, just saying...
            val deferredOne = async { appRepository.mainEntity() }
            val deferredTwo = async { appRepository.mainEntityAdditional() }

            var resultOne: List<MainEntity>? = deferredOne.await().data
            var resultTwo: List<MainEntity>? = deferredTwo.await().data

            resultOne?.let { finalList = finalList.union(resultOne).toList() }
            resultTwo?.let { finalList = finalList.union(resultTwo).toList() }

        }

        val result = Resource.success(finalList)

        return result


    }


    class RequestValues : Interactor.RequestValues// in this case we don't need nothing for this use case
}
